File: packages/php/src/Seguranca/Ataques/AtaqueLengthExtension.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\CriptografiaAlvo;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Length extension / truncamento: a estrutura do token é
 * integridade[32] . ciphertext[n] . iv[16], e o MAC cobre iv+ciphertext.
 * Truncar, estender ou deslocar bytes não pode produzir um token aceito.
 */
class AtaqueLengthExtension implements AtaqueInterface
{
    public function __construct(private string $mensagem = 'texto para o ataque de length extension')
    {
    }

    public function nome(): string
    {
        return 'Length extension / truncamento de token';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        if (!$alvo instanceof CriptografiaAlvo) {
            throw new SkipAtaqueException('Precisa de base64url_encode/decode do alvo.');
        }

        $prefixo = $alvo->prefixo();
        $token = $alvo->encrypt($this->mensagem);
        $corpo = substr($token, strlen($prefixo));
        $meio = intdiv(strlen($corpo), 2);

        $variantes = [
            'append A' => $prefixo . $corpo . 'A',
            'append =' => $prefixo . $corpo . '=',
            'truncar 1 char' => $prefixo . substr($corpo, 0, -1),
            'truncar 2 chars' => $prefixo . substr($corpo, 0, -2),
            'inserir ! no meio' => $prefixo . substr($corpo, 0, $meio) . '!' . substr($corpo, $meio),
            'prefixo extra' => $prefixo . 'A' . $corpo,
        ];

        // variantes que mexem nos campos decodificados
        try {
            $decoded = $alvo->base64urlDecode($corpo);
            $campos = $alvo->decompor($decoded);

            $ctMaior = $campos['ciphertext'] . chr(0x41);
            $variantes['ciphertext +1 byte'] = $prefixo . $alvo->base64urlEncode($alvo->recompor([
                'integridade' => $campos['integridade'],
                'ciphertext' => $ctMaior,
                'iv' => $campos['iv'],
            ]));

            $ctMenor = substr($campos['ciphertext'], 0, max(0, strlen($campos['ciphertext']) - 1));
            $variantes['ciphertext -1 byte'] = $prefixo . $alvo->base64urlEncode($alvo->recompor([
                'integridade' => $campos['integridade'],
                'ciphertext' => $ctMenor,
                'iv' => $campos['iv'],
            ]));

            $ivMaior = $campos['iv'] . chr(0x42);
            $variantes['iv +1 byte'] = $prefixo . $alvo->base64urlEncode($alvo->recompor([
                'integridade' => $campos['integridade'],
                'ciphertext' => $campos['ciphertext'],
                'iv' => $ivMaior,
            ]));
        } catch (\Throwable $e) {
            // se a decodificação falhar, segue com as variantes textuais
        }

        $aceitas = [];
        foreach ($variantes as $nome => $v) {
            if ($v === $token) {
                continue;
            }
            try {
                $alvo->decrypt($v);
                $aceitas[] = $nome;
            } catch (\Throwable $e) {
                // esperado
            }
        }

        $vulneravel = !empty($aceitas);

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'critica' : 'info',
            detalhes: $vulneravel
                ? count($aceitas) . ' variante(s) aceita(s): ' . implode('; ', $aceitas)
                : 'Todas as ' . count($variantes) . ' variantes de truncamento/extensão foram rejeitadas',
            dados: ['aceitas' => $aceitas],
        );
    }
}

File: packages/rust/src/seguranca/ataques/ataque_length_extension.rs
//! Length extension / truncamento: a estrutura do token é
//! integridade[32] . ciphertext[n] . iv[16], e o MAC cobre iv+ciphertext.
//! Truncar, estender ou deslocar bytes não pode produzir um token aceito.

use crate::seguranca::alvo_criptografico::{AlvoCriptografico, CamposToken};
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};

pub struct AtaqueLengthExtension {
    pub mensagem: String,
}

impl AtaqueLengthExtension {
    pub fn new() -> Self {
        AtaqueLengthExtension {
            mensagem: "texto para o ataque de length extension".to_string(),
        }
    }
}

impl Default for AtaqueLengthExtension {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueLengthExtension {
    fn nome(&self) -> String {
        "Length extension / truncamento de token".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let prefixo = alvo.prefixo();
        let token = alvo
            .encrypt(self.mensagem.as_bytes())
            .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
        let corpo = &token[prefixo.len()..];
        let meio = corpo.len() / 2;

        let mut variantes: Vec<(String, String)> = Vec::new();
        variantes.push((
            "append A".to_string(),
            format!("{}{}A", prefixo, corpo),
        ));
        variantes.push((
            "append =".to_string(),
            format!("{}{}=", prefixo, corpo),
        ));
        variantes.push((
            "truncar 1 char".to_string(),
            format!("{}{}", prefixo, &corpo[..corpo.len() - 1]),
        ));
        variantes.push((
            "truncar 2 chars".to_string(),
            format!("{}{}", prefixo, &corpo[..corpo.len() - 2]),
        ));
        variantes.push((
            "inserir ! no meio".to_string(),
            format!("{}{}!{}", prefixo, &corpo[..meio], &corpo[meio..]),
        ));
        variantes.push((
            "prefixo extra".to_string(),
            format!("{}A{}", prefixo, corpo),
        ));

        // variantes que mexem nos campos decodificados
        if let Ok(decoded) = alvo.base64url_decode(corpo) {
            let campos = alvo.decompor(&decoded);

            let mut ct_maior = campos.ciphertext.clone();
            ct_maior.push(0x41);
            let campos_ct_maior = CamposToken {
                integridade: campos.integridade.clone(),
                ciphertext: ct_maior,
                iv: campos.iv.clone(),
            };
            variantes.push((
                "ciphertext +1 byte".to_string(),
                format!(
                    "{}{}",
                    prefixo,
                    alvo.base64url_encode(&alvo.recompor(&campos_ct_maior))
                ),
            ));

            let fim = campos.ciphertext.len().saturating_sub(1);
            let ct_menor = campos.ciphertext[..fim].to_vec();
            let campos_ct_menor = CamposToken {
                integridade: campos.integridade.clone(),
                ciphertext: ct_menor,
                iv: campos.iv.clone(),
            };
            variantes.push((
                "ciphertext -1 byte".to_string(),
                format!(
                    "{}{}",
                    prefixo,
                    alvo.base64url_encode(&alvo.recompor(&campos_ct_menor))
                ),
            ));

            let mut iv_maior = campos.iv.clone();
            iv_maior.push(0x42);
            let campos_iv_maior = CamposToken {
                integridade: campos.integridade.clone(),
                ciphertext: campos.ciphertext.clone(),
                iv: iv_maior,
            };
            variantes.push((
                "iv +1 byte".to_string(),
                format!(
                    "{}{}",
                    prefixo,
                    alvo.base64url_encode(&alvo.recompor(&campos_iv_maior))
                ),
            ));
        }

        let mut aceitas: Vec<String> = Vec::new();
        for (nome, v) in &variantes {
            if v == &token {
                continue;
            }
            if alvo.decrypt(v).is_ok() {
                aceitas.push(nome.clone());
            }
        }

        let vulneravel = !aceitas.is_empty();

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Critica
            } else {
                Severidade::Info
            },
            &if vulneravel {
                format!(
                    "{} variante(s) aceita(s): {}",
                    aceitas.len(),
                    aceitas.join("; ")
                )
            } else {
                format!(
                    "Todas as {} variantes de truncamento/extensão foram rejeitadas",
                    variantes.len()
                )
            },
        ))
    }
}
